### PR TITLE
Simplify the trait bounds, fixes #54

### LINKED ISSUE #54: hkdf: simplify trait bounds
Initially reported in [Zulip](https://rustcrypto.zulipchat.com/#narrow/stream/260043-KDFs/topic/HKDF.20trait.20bounds).

Right now if function uses HKDF and wants to be generic over hash functions, trait bounds look quite scary:
```rust
fn hkdf<H: CoreProxy>()
where
    <H as CoreProxy>::Core: HashMarker
        + UpdateCore
        + FixedOutputCore
        + BufferKindUser<BufferKind = Eager>
        + Default
        + Clone,
    <<H as CoreProxy>::Core as BlockSizeUser>::BlockSize: IsLess<U256>,
    Le<<<H as CoreProxy>::Core as BlockSizeUser>::BlockSize, U256>: NonZero,
{
    Hkdf::<H>::new(None, &[]);
}
```

It could be worth to simplify the bounds somehow. Also it may be worth to allow `hkdf` to use the [`SimpleHmac`](https://docs.rs/hmac/latest/hmac/struct.SimpleHmac.html) in addition to [`Hmac`](https://docs.rs/hmac/latest/hmac/type.Hmac.html). It probably will involve introducing a helper trait or two in `hmac`.

### PR DESCRIPTION
**Pull request by jbis9051**
Fixes #54

### CODE REVIEW

**Comment by jbis9051** [comment]
I replaced Hmac with SimpleHmac. An alternative would be to create a SimpleHkdf.

Using SimpleHmac is preferable because it will allow hashes like the BLAKE family to be used. See: https://github.com/BLAKE3-team/BLAKE3/issues/223

File: hkdf/src/lib.rs [hunk 1]
```diff
@@ -167,62 +142,28 @@ where
...
-        + Default
-        + Clone,
-    <D::Core as BlockSizeUser>::BlockSize: IsLess<U256>,
-    Le<<D::Core as BlockSizeUser>::BlockSize, U256>: NonZero,
-{
+impl<D: Digest + BlockSizeUser + fmt::Debug> fmt::Debug for HkdfExtract<D> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.write_str("HkdfExtract<")?;
-        <D::Core as AlgorithmName>::write_alg_name(f)?;
+        D::fmt(&D::new(), f)?;
```
Comment (jbis9051) [inline hkdf/src/lib.rs:155]: Is there a better way to do this?

Comment (tarcieri) [reply hkdf/src/lib.rs:155]: I'd probably replace the whole thing with a `write!` directive or use `DebugStruct`:

```suggestion
        write!(f, "HkdfExtract<{:?}> { ... } ", &D::new())
```

**Comment by jbis9051** [comment]
Wonder if I can make it generic over Hmac implementation. The user could provide Hmac or SimpleHmac depending on what they need. We would reexport Hmac and SimpleHmac.

File: hkdf/src/lib.rs [hunk 2]
```diff
@@ -167,62 +142,28 @@ where
...
-        + Default
-        + Clone,
-    <D::Core as BlockSizeUser>::BlockSize: IsLess<U256>,
-    Le<<D::Core as BlockSizeUser>::BlockSize, U256>: NonZero,
-{
+impl<D: Digest + BlockSizeUser + fmt::Debug> fmt::Debug for HkdfExtract<D> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.write_str("HkdfExtract<")?;
-        <D::Core as AlgorithmName>::write_alg_name(f)?;
+        write!(f, "HkdfExtract<{:?}> {{ ... }}", &D::new())?;
```
Comment (tarcieri) [inline hkdf/src/lib.rs:155]: You can get rid of the `f.write_str` above and below this, since it's all covered by the `write!` now

**Comment by jbis9051** [comment]
I'd like to ideally get the interface to `Hkdf<Sha256, SimpleHmac>`. I've currently (not pushed yet) gotten it to `Hkdf<Sha256, SimpleHmac<Sha256>>`.

**Comment by jbis9051** [comment]
So this is another option. We make the Hmac implementation generic so the user can provide SimpleHmac or Hmac depending on the algo they used. I think it's better than simply replacing Hmac with SimpleHmac.  

@tarcieri if you like this approach, I can add a couple tests for testing both Hmac and SimpleHmac. I could also either reexport Hmac and SimpleHmac or remove the dependency. 

A couple considerations:

Do we really want users to be able to pass anything to Hkdf? Maybe we should add a dummy trait to restrict passing either Hmac<D> or SimpleHmac<D> as there maybe other things that would fulfills the trait requirements but wouldn't be appropriate.

**Comment by newpavlov** [comment]
#57 has been merged, so I will close this PR. But anyway thank you for it!

**Pull request closed without merging**